sqlalchemy_bigquery/base.py

```diff
@@ -686,6 +771,147 @@ def visit_drop_table_comment(self, drop):
+            )
+
+        return True
+
+    def _raise_for_type(self, option, value, expected_type):
+        if type(value) is not expected_type:
+            raise TypeError(
+                f"bigquery_{option} dialect option accepts only {expected_type},"
+                f" provided {repr(value)}"
+            )
+
+    def _process_time_partitioning(
+        self, table: Table, time_partitioning: TimePartitioning
+    ):
+        """
+        Generates a SQL 'PARTITION BY' clause for partitioning a table by a date or timestamp.
+
+        Args:
+        - table (Table): The SQLAlchemy table object representing the BigQuery table to be partitioned.
+        - time_partitioning (TimePartitioning): The time partitioning details,
+            including the field to be used for partitioning.
+
+        Returns:
+        - str: A SQL 'PARTITION BY' clause that uses either TIMESTAMP_TRUNC or DATE_TRUNC to
+        partition data on the specified field.
+
+        Example:
+        - Given a table with a TIMESTAMP type column 'event_timestamp' and setting
+        'time_partitioning.field' to 'event_timestamp', the function returns
+        "PARTITION BY TIMESTAMP_TRUNC(event_timestamp, DAY)".
+        """
+        field = "_PARTITIONDATE"
+        trunc_fn = "DATE_TRUNC"
+
+        if time_partitioning.field is not None:
+            field = time_partitioning.field
+            if isinstance(
+                table.columns[time_partitioning.field].type,
+                sqlalchemy.sql.sqltypes.TIMESTAMP,
+            ):
+                trunc_fn = "TIMESTAMP_TRUNC"
+
+        return f"PARTITION BY {trunc_fn}({field}, {time_partitioning.type_})"
+
+    def _process_range_partitioning(
+        self, table: Table, range_partitioning: RangePartitioning
+    ):
+        """
+        Generates a SQL 'PARTITION BY' clause for partitioning a table by a range of integers.
+
+        Args:
+        - table (Table): The SQLAlchemy table object representing the BigQuery table to be partitioned.
+        - range_partitioning (RangePartitioning): The RangePartitioning object containing the
+        partitioning field, range start, range end, and interval.
+
+        Returns:
+        - str: A SQL string for range partitioning using RANGE_BUCKET and GENERATE_ARRAY functions.
+
+        Raises:
+        - ValueError: If the partitioning field is not defined, not an integer type, or if the range
+        start/end values are not integers.
+
+        Example:
+            "PARTITION BY RANGE_BUCKET(zipcode, GENERATE_ARRAY(0, 100000, 10))"
+        """
+        if range_partitioning.field is None:
+            raise ValueError("bigquery_range_partitioning expects field to be defined")
+
+        if not isinstance(
+            table.columns[range_partitioning.field].type,
+            sqlalchemy.sql.sqltypes.INT,
+        ):
+            raise ValueError(
+                "bigquery_range_partitioning expects field data type to be INTEGER"
+            )
```

Comment:
```suggestion
            raise TypeError(
                "bigquery_range_partitioning expects field data type to be INTEGER"
            )
```

I feel like we are raising the wrong type of error here.
```
TypeError: Raised when an operation or function is applied to an object of inappropriate type.
```

```
ValueError: Raised when an operation or function receives an argument that has
the right type but an inappropriate value...
```

I have not checked yet to see if any tests look for a specific kind of error in this location. IF yes, those tests will need an update, as well.

Reply:
I see what you are saying. How about the above verbiage change to the error message to help future me remember what this error means?

Reply:
Ok!